services/skus/handler/handler.go

```diff
@@ -0,0 +1,81 @@
+package handler
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	"github.com/asaskevich/govalidator"
+
+	"github.com/brave-intl/bat-go/libs/handlers"
+	"github.com/brave-intl/bat-go/libs/logging"
+	"github.com/brave-intl/bat-go/libs/requestutils"
+
+	"github.com/brave-intl/bat-go/services/skus/model"
+)
+
+type orderService interface {
+	CreateOrderFromRequest(ctx context.Context, req CreateOrderRequest) (*model.Order, error)
+}
+
+type Order struct {
+	svc orderService
+}
+
+func NewOrder(svc orderService) *Order {
+	result := &Order{
+		svc: svc,
+	}
+
+	return result
+}
+
+func (h *Order) Create(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+	ctx := r.Context()
+
+	var req CreateOrderRequest
+	if err := requestutils.ReadJSON(r.Context(), r.Body, &req); err != nil {
+		return handlers.WrapError(err, "Error in request body", http.StatusBadRequest)
+	}
+
+	if _, err := govalidator.ValidateStruct(req); err != nil {
+		return handlers.WrapValidationError(err)
+	}
```

Comment:
If you look at the `valid` struct tags on the request types, you'd realise that this call is absolutely useless:
1. Most of the fields have their validation disabled (via `-`).
2. The only field that requires validation is `Quantity`, but it specifies `int`, which does not make much sense since. Validation takes place after unmarshalling, and by the time validation is called, it's already proved to be a number. So it's useless.

After reviewers have seen this comment, I will remove this call.

PS. Also, in my experience, I preferred the other [validator](https://github.com/go-playground/validator) over this one. But that's OK for now.

Reply:
Why is the validation disabled, do we not need to check certain fields such as `email` or `sku` are required i.e. **`valid:"required"`**?

Arguably, you could put some sanity checks in at this point as well i.e. that `quantity` is not only an int but also a positive int or even within a specified range? I don't believe that validation is currently happening in the domain.

I think a lot the disabled validation is due to workarounds from [this](https://github.com/brave-intl/bat-go/blob/7c52439d84acadd18bae3602893bc06c4a88bac7/services/grant/cmd/grant.go#L288) and [this](https://github.com/asaskevich/govalidator/blob/a9d515a09cc289c60d55064edec5ef189859f172/validator.go#L37)

Would it be beneficial to improve the validation rather than remove? It should be safe enough as the client should be erroring anyway given this is bad data.

Reply:
I agree that having validation is important and desired, and I am eventually planning on enabling it. However, I don't plan on doing it as part of this PR, and the comment was made in hopes to understand why it had been disabled in the first place.

Generally, it's a good idea for a refactor to not change behaviour. That's why I am not removing (since it's disabled) nor enabling validation here.

When it comes to enabling, as outlined in the comment, I will propose an alternative solution a bit later.

Reply:
As for the reasons behind my desire to change the validation library (when will be addressing the validation question later), they are straightforward:
1. At some point, I'd spent some time choosing a validation library out of the several available, and came to the conclusion that [GoPlayground Validator](https://github.com/go-playground/validator) was superior overall.
2. [Validator](https://github.com/go-playground/validator) is more powerful than [Govalidator](https://github.com/asaskevich/govalidator) that is being used currently.
3. [Validator](https://github.com/go-playground/validator) is by and large faster and more resource efficient, by an order and more times; in terms of memory – even more.

To prove the last point, one can do a simple benchmark for the most trivial (and oft-seen) use case – validating struct fields. Let's say we want to validate a struct with a string and an int fields.

Here's benchmark code:

```golang
package main

import (
	"testing"

	"github.com/asaskevich/govalidator"
	"github.com/go-playground/validator/v10"
)

func BenchmarkStructValidator(b *testing.B) {
	validate := validator.New()
	type Foo struct {
		StringValue string `validate:"min=5,max=10"`
		IntValue    int    `validate:"min=5,max=10"`
	}

	validFoo := &Foo{StringValue: "Foobar", IntValue: 7}

	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		_ = validate.Struct(validFoo)
	}
}

func BenchmarkStructGoValidator(b *testing.B) {
	type Foo struct {
		StringValue string `valid:"minstringlength(5),maxstringlength(10)"`
		IntValue    int    `valid:"range(5|10)"`
	}

	validFoo := &Foo{StringValue: "Foobar", IntValue: 7}

	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		_, _ = govalidator.ValidateStruct(validFoo)
	}
}
```

Here's how I ran it, and results:

```bash
$ go test -benchmem -benchtime 5s -bench=Benchmark -count 8 -timeout 600s -cpu=8 | tee bench.out
goos: darwin
goarch: arm64

BenchmarkStructValidator-8     	52194601	       114.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructValidator-8     	51983551	       114.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructValidator-8     	52311400	       114.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructValidator-8     	52233678	       114.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructValidator-8     	52238187	       114.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructValidator-8     	52211179	       114.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructValidator-8     	52066120	       114.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructValidator-8     	52107739	       114.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructGoValidator-8   	 1660989	      3607 ns/op	    1770 B/op	      37 allocs/op
BenchmarkStructGoValidator-8   	 1661774	      3602 ns/op	    1768 B/op	      37 allocs/op
BenchmarkStructGoValidator-8   	 1664119	      3599 ns/op	    1767 B/op	      37 allocs/op
BenchmarkStructGoValidator-8   	 1666016	      3609 ns/op	    1769 B/op	      37 allocs/op
BenchmarkStructGoValidator-8   	 1662258	      3607 ns/op	    1769 B/op	      37 allocs/op
BenchmarkStructGoValidator-8   	 1663791	      3604 ns/op	    1769 B/op	      37 allocs/op
BenchmarkStructGoValidator-8   	 1661024	      3611 ns/op	    1769 B/op	      37 allocs/op
BenchmarkStructGoValidator-8   	 1662559	      3612 ns/op	    1767 B/op	      37 allocs/op
```

Organised results:

```bash
benchstat bench.out
goos: darwin
goarch: arm64

                    │   sec/op    │
StructValidator-8     114.1n ± 0%
StructGoValidator-8   3.607µ ± 0%

                    │      B/op      │
StructValidator-8       0.000 ± 0%
StructGoValidator-8   1.728Ki ± 0%

                    │  allocs/op   │
StructValidator-8     0.000 ± 0%
StructGoValidator-8   37.00 ± 0%
```

The results speak for themselves, the GoPlayground Validator is miles ahead of Govalidator.